### PR TITLE
Improved UI for handling many sources

### PR DESCRIPTION
**Pull request by tpietzsch**
Builds on PR #67, supersedes PR #65 

Based on initial work by @tibuch, this introduces a collapsible "card panel" on the right side of the BigDataViewer window and "cards" for handling BDV sources and groups.

Card Panel
-----------
The card panel can be expanded/collapse by moving the mouse to the right side of the window and clicking the arrow that appears. Cards can be collapsed/expanded and resized.
splitpanel

Short-cuts (configurable):

| in ViewerPanel | |
|---|---|
|`P` | focus card panel (and expand it if it is collaped). |
|`shift P`, `shift ESC` | collapse card panel. |

| in CardPanel | |
|---|---|
| `ESC` | focus ViewerPanel |
| `shift ESC` | collapse card panel |

Sources Card
--------------
The "Sources" card allows to make sources "active" (visible in "fused" display mode) or "current" (visible in "single-source" display mode). When clicking on the "active" checkbox of one of multiple selected sources, all of them are toggled "active".
activecurrent

Short-cuts (configurable):

| in Sources table | |
|---|---|
| `A` | toggle activeness of selected sources |
| `C` | cycle currentness among selected sources |

plus the usual table shortcuts (ctrl/cmd-A to select all, shift click for range selection, etc).

Below the table is a panel where the color/min/max of the selected sources can be changed.
colors

When selecting a set of sources which do not have the same color and/or display range, the background of the respective panel turns red to indicate this.

The range on the slider can be configured by a context menu (per set of sources). Manually entering min/max numbers that are out of range, automatically expands the range.

Colors can also be changed by clicking directly on the color fields in the table.

Groups Card
-------------
In the "Groups" card, groups can be created/expanded by drag-and-dropping sources into the tree.
Sources can be removed from groups, or whole groups can be removed, by selecting them in the tree and pressing `DELETE` or `BACK_SPACE`.
grouping

Below the groups tree, is a panel where the color/min/max of the sources in the selected groups can be changed. It is identical to the one in the "Sources" card.
groupcolors

The "Groups" tree allows to make groups "active" (visible in "fused group" display mode) or "current" (visible in "single-group" display mode). When clicking on the "active" checkbox of one of multiple selected groups, all of them are toggled "active".

Note that a source can be in multiple groups, allowing to toggle overlapping subsets in/visible, which is really powerful.

Short-cuts (configurable):

| in Groups tree | |
|---|---|
| `A` | toggle activeness of selected groups |
| `C` | cycle currentness among selected groups |
| `BACK_SPACE`, `DELETE` | remove selected groups, and remove selected sources from containing group |
| ENTER | start editing name of selected group |

Future work
------------

* Extensibility. @tibuch and @frauzufall are already working on this. Basically you can annotate you own card plugins with scijava annotations and they will be discovered and shown if all the right preconditions are satisfied. This will probably first appear in vistools, then maybe later in bigdataviewer-core.
* A "Display Mode" card having 
image and interpolation mode.
* A way to reorder groups.
* An extensible way to sort the sources table, such that you can bring sources that you probably want to group next to each other.
* There are probably minor synchronization issues left when using the UI while doing programmatic modifications. If you ever come across a dead-lock while using this, please, please, please, attach visualVM, do a thread-dump, and send it to me.
* There are probably hard-to-reproduce UI glitches. I think I have caught a lot of them, but I'm sure some remain. If you come across them please report, preferably with instructions how to reproduce.
* If people hate it much, I will add a way to turn it off by default.

If you care deeply / want to work on any of these, please create issues.

### CODE REVIEW

**Comment by tpietzsch** [comment]
> Future work

* Update bigdataviewer-vistools to use the new `ViewerState` API exclusively.
* A mode where (only) the selected sources/groups are visible (a la BigStitcher). E.g., by automatically switching display mode and toggling activeness).

File: pom.xml [hunk 1]
```diff
@@ -139,6 +139,7 @@
 		<imglib2.version>5.8.0</imglib2.version>
 		<imglib2-cache.version>1.0.0-beta-12</imglib2-cache.version>
 		<spim_data.version>2.2.3</spim_data.version>
+		<ui-behaviour.version>2.0.1</ui-behaviour.version>
```
Comment (frauzufall) [inline pom.xml]: If I don't add this line to my own project, I get a `NoSuchMethodError: org.scijava.ui.behaviour.util.InputActionBindings.installNewBindings`. We need a PR to `pom-scijava` with these new versions.

**Comment by bogovicj** [comment]
Ran this test on a workstation with running ubuntu 16.04. Here are some initial thoughts

### Old dialogs

The old "brightness/contrast" and "visibility and grouping" dialogs are still around. Should we :

1) Get rid of it now?
2) Deprecate and remove in the future? 
3) Plan to keep around indefinitely?

### State indicators?

Some aspects of the state of the viewer which were explicitly indicated in the old "visibility and grouping" dialog are not explicit in the new card.  Specifically:

1) The sources that are currently visible
2) Whether "fused" mode is on
3) Whether "grouping" mode is on

Maybe add another card called "State" or something like that.  Or to have the `Sources` and `Groups` banners highlighted when active. 
                                                                                                                                               
### Renameable sources? (not so important)

I know that source names are final [for some implementations](https://github.com/bigdataviewer/bigdataviewer-core/blob/master/src/main/java/bdv/AbstractSpimSource.java#L121), but do they have to be?

**Comment by tpietzsch** [comment]
>   * improvement: add right click option (I will add that)
👍 

> * dragging sources at the end of group list creates new group -> this is difficult to know
>   * improvement: start with zero groups and note that dragging sources there will create a group; this needs to wait for backwards compatibility reasons

We could add logic to display this label "that dragging sources there will create a group" when there are no groups. The "start with zero groups" should be fine when used through bdv-vistools. The default backwards compatibility behaviour is only necessary when starting BigDataViewer directly I think ... (will have to look into it)

Anyway,  a button / context menu item "create new group" should be added.

> * while in transform mode (T), changing the active source does not change the source which is transformed

This never worked and is not connected to this PR
>   * improvement: freeze changing the active source while in transform mode or actually change active source of transform action

That would be a useful addition.

> * in Linux, editing a group name makes the group entry disappear (I will look into that)
Thanks!

> * when settings the bounds, I would expect the min/max markers to also change accordingly (e.g. if the markers are at 0..1 and I change to bounds 0..255, I would have though the markers would go to 0..255 too)
They should and for me they do. Let's look at this together

> 
> ### Issues with current vistools master branch
> * coloring, value range, visibility of groups does not work (I will work on that)

Yes, vistools is not adapted yet. I'll work on that.

> * we should add a `DiscoverableCard` interface (`getComponent`, `getName`, `getVisibility`) -> I will make a PR and add a static method to `vistools` for creating and adding discoverable cards

👍

**Comment by tpietzsch** [comment]
> ### Old dialogs
> The old "brightness/contrast" and "visibility and grouping" dialogs are still around. Should we :
> 
> 1. Get rid of it now?
> 2. Deprecate and remove in the future?
> 3. Plan to keep around indefinitely?

I would do 2.
Or 3. and disable them by default, if some people still prefer them.

> 
> ### State indicators?
> Some aspects of the state of the viewer which were explicitly indicated in the old "visibility and grouping" dialog are not explicit in the new card. Specifically:
> 
> 1. The sources that are currently visible
> 2. Whether "fused" mode is on
> 3. Whether "grouping" mode is on
> 
> Maybe add another card called "State" or something like that. Or to have the `Sources` and `Groups` banners highlighted when active.

Yes, definitely.
@tibuch is already on it. (https://github.com/bigdataviewer/bigdataviewer-core/pull/70)

> 
> ### Renameable sources? (not so important)

Yes, I agree, not so important now. But would be a nice future addition.

**Comment by tischi** [comment]
Could there be (is there already) an option to generally disable any UI (both the old and the new one)? 
Such that the bdv could be used as a just as viewer within applications/code that have their own UI for the displayed `SourceAndConverter`s.

**Comment by frauzufall** [comment]
@tischi The cards are in the `ViewerFrame`, so if you use `BdvHandlePanel` to display your stuff or the vistools method `BdvFunctions.show(img, name, Bdv.options().addTo(bdvHandle));`, the side panel will not exist.

**Comment by tpietzsch** [comment]
@tischi you can just use the stuff in `bigdataviewer-core`. For example, create your own `ViewerPanel`. That has nothing (except a time slider). Have a look at `ViewerFrame` and `BigDataViewer` classes. They add all the surrounding dialogs, menus, ... Just pick what you like and leave away the rest.

**Comment by NicoKiaru** [comment]
I noticed a small glitch with Manual Transform: in single-source mode, pressing `T` lead to a warning message : "Can only do manual transform when in FUSED mode ". However pressing a second time `T`, still in single-source mode, lead to the message "Fixed manual transform" and to a change of view transform which is quite weird, at 45 degrees.

**Comment by tpietzsch** [comment]
@NicoKiaru Thanks for reporting! This doesn't seem to be related to the new UI. It also happens in BigDataViewer master. I made a new issue #75 for it.

**Comment by tpietzsch** [comment]
@NicoKiaru It was easy to fix (#76).

**Comment by NicoKiaru** [comment]
I experienced a lock when trying to add a `SourceAndConverter` programmatically to bdv:

https://github.com/bigdataviewer/bigdataviewer-core/issues/77

All what I could dig is in the issue.

Hope that helps!

**Comment by tpietzsch** [comment]
> Hope that helps!

Yes, very much! I fixed the issue.

**Comment by NicoKiaru** [comment]
Hi all, what's the programmatic way to make a source(AndConverter) visible now, since `getVisibilityAndGrouping()` is deprecated ?

**Comment by NicoKiaru** [comment]
Just mentioning here this new thread deadlock:  https://github.com/bigdataviewer/bigdataviewer-core/issues/78

**Comment by tpietzsch** [comment]
> Just mentioning here this new thread deadlock: #78

Ok, it looks like there might be more of those deadlocks hiding. I will go through everything again and try to fix as many as possible. Basically this should involve:
1) UI modifications only on the AWT/Swing thread. This is anyway how it should be done, but I have been lazy about it...
2) Losen the synchronization on `SynchronizedViewerState`. Ideally, only the modifying methods need to be `synchronized`. However, this is only possible if it can be guaranteed that all the reading methods provide a consistent result of either the state before concurrent modification or after, not some kind of weird in-between state. I need to check whether that is possible.

I need a stretch of focused time working on this. I'll try to do it some day next week.

Thanks for reporting!

**Comment by NicoKiaru** [comment]
Hi! Do you know how difficult / when  / if it is be possible to make `BdvHandle` work with the new UI ?

Right now me and @tischi are using this class heavily. I've tried the new UI, but I had to change all `BdvHandle` instances by `BigDataViewer` instances, and this is tedious, so we'll probably wait to have BdvHandle working with the new UI in order to adapt our repo.

**Comment by tpietzsch** [comment]
> Hi! Do you know how difficult / when / if it is be possible to make `BdvHandle` work with the new UI ?

Yes, of course, I have been working on this. It is ready and should be released soon, you can find it on this branch: https://github.com/bigdataviewer/bigdataviewer-vistools/tree/table-tree-cards

**Pull request merged**